functional_programming_tests/test/without_functional_programming.cpp

```diff
@@ -0,0 +1,205 @@
+			return std::nullopt;
+		}            
+		path.push_back({path.back().x, path.back().y+del_y_sign});
+	}
+
+	return path;
+}
+
+// Function that gets costmap
+void PathGenerator::costmap_callback(const std_msgs::msg::UInt8MultiArray::SharedPtr msg) { // Action
+    // Get the costmap from a ros topic
+	// Check that map layout makes sense
+	if ((msg->layout.dim[0].size * msg->layout.dim[1].size) != msg->layout.dim[0].stride) {
+		RCLCPP_ERROR_STREAM(node_->get_logger(), "COSTMAP DIMENSIONS AND STRIDE INCONSISTENT!!");
+		return;
+	}
+	if (msg->layout.dim[0].stride != msg->data.size()) {
+		RCLCPP_ERROR_STREAM(node_->get_logger(), "COSTMAP LENGTH AND STRIDE INCONSISTENT!!");
+		return;
+	}
+	auto const begin = std::begin(msg->data) ;
+	// Populate the map
+	for (auto row = 0; row < msg->layout.dim[0].size; row++) {
+		map_.push_back({begin+row*msg->layout.dim[1].size, begin+(row+1)*msg->layout.dim[1].size});
+	}
+}
+
+class TaskPlanningFixture : public testing::Test {
+public:
+	// Adapted from minimal_integration_test
+	TaskPlanningFixture() : node_(std::make_shared<rclcpp::Node>("bad_test_publisher")), executor_(std::make_shared<rclcpp::executors::SingleThreadedExecutor>()) {
+        // Create ROS2 publisher for the costmap
+		costmap_publisher_ = node_->create_publisher<std_msgs::msg::UInt8MultiArray>("test/costmap",1);
+
+        // Publish the costmap every 100 ms
+		timer_ = node_->create_wall_timer(100ms, std::bind(&TaskPlanningFixture::publish_costmap, this));
+	}
+
+	void SetUp() override {
+		executor_->add_node(node_);
+		executor_thread_ = std::thread([this]() { executor_->spin(); });    
+	}
+
+    // Cleanup actions that could throw an exception
+	void TearDown() override {
+		executor_->cancel();
+		executor_thread_.join();
+	}
+
+protected:
+	void publish_costmap() { // Action
+		auto msg = std_msgs::msg::UInt8MultiArray();
+
+		msg.layout.dim.resize(3, std_msgs::msg::MultiArrayDimension());
+
+	    msg.layout.dim[0].label = "rows";
+	    msg.layout.dim[0].size = 8;
+	    msg.layout.dim[0].stride = 64;
+
+	    msg.layout.dim[1].label = "columns";
+	    msg.layout.dim[1].size = 8;
+	    msg.layout.dim[1].stride = 1;
+
+	    msg.layout.dim[2].label = "channel";
+	    msg.layout.dim[2].size = 1;
+	    msg.layout.dim[2].stride = 1;    
+
+		msg.data = {0, 0, 0, 0, 0, 0, 0, 0, 
+		0, 0, 0, 1, 0, 0, 0, 0, 
+		0, 0, 0, 1, 0, 0, 0, 0, 
+		0, 0, 1, 1, 1, 0, 0, 0, 
+		0, 0, 1, 0, 1, 1, 0, 0, 
+		0, 0, 1, 0, 0, 0, 0, 0, 
+		0, 0, 0, 0, 0, 0, 0, 0, 
+		0, 0, 0, 0, 0, 0, 0, 0}; 
+
+		costmap_publisher_->publish(msg);
+	};
+	rclcpp::Node::SharedPtr node_;
+	rclcpp::Executor::SharedPtr executor_;
+	std::thread executor_thread_;
+	rclcpp::TimerBase::SharedPtr timer_;
+
+	PathGenerator pg_;
+
+	rclcpp::Publisher<std_msgs::msg::UInt8MultiArray>::SharedPtr costmap_publisher_;
+};
+
+TEST_F(TaskPlanningFixture, same_start_and_goal) {
+	// Wait some time to make sure a valid costmap has been received
```

Comment:
Note how this becomes a source of bugs and at a minimum slows down your tests

Reply:
Done